packages/wonder-stuff-core/src/kind-error.js

```diff
@@ -0,0 +1,125 @@
+     * the stack references the point at which an error is thrown which can
+     * be useful when helper functions are used to build the error being thrown.
+     * Defaults to `0`.
+     * @memberof KindError
+     */
+    constructor(
+        message: string,
+        kind: string = Errors.Unknown,
+        {
+            cause,
+            prefix,
+            metadata,
+            stripStackFrames,
+            minimumFrameCount,
+        }: Options = {
+            cause: undefined,
+            metadata: undefined,
+            prefix: "",
+            stripStackFrames: 0,
+            minimumFrameCount: 1,
+        },
+    ) {
+        // Validate arguments.
+        if (cause && !(cause instanceof Error)) {
+            throw new Error("cause must be an instance of Error");
+        }
+        if (kind && /\s/g.test(kind)) {
+            throw new Error("kind must not contain whitespace");
+        }
+        if (prefix && /\s/g.test(prefix)) {
+            throw new Error("prefix must not contain whitespace");
+        }
+        if (stripStackFrames && stripStackFrames < 0) {
+            throw new Error("stripStackFrames must be >= 0");
+        }
+        if (minimumFrameCount && minimumFrameCount < 0) {
+            throw new Error("minimumFrameCount must be >= 0");
+        }
+
+        super(message);
+
+        // Save the original message as we may change this message later.
+        this.originalMessage = message;
+
+        // Metadata associated with this error.
+        this.metadata = cloneMetadata(metadata);
+
+        // Set the name so we get a nice error output, like
+        // KAInternalError
+        this.name = `${prefix ?? ""}${kind}Error`;
+
+        // The kind of error which can be used for grouping with
+        // other error sources that use the same error taxonomy.
+        this.kind = kind;
+
+        if (cause != null) {
+            // We want to generate a better stack trace using the source error
+            // and our own stack.
+            const normalizedError = ErrorInfo.normalize(
+                this,
+                stripStackFrames,
+                minimumFrameCount,
+            );
+            const normalizedCause = ErrorInfo.normalize(cause);
+            const combined = ErrorInfo.combine(
+                normalizedError,
+                normalizedCause,
+            );
+
+            /**
+             * $FlowIgnore[incompatible-type]
+             * Flow doesn't like us leaving KAError temporarily without a
+             * stack, but I don't want this crashing because we try to
+             * assign to a readonly thing, so being mildly cautious.
+             */
+            delete this.stack;
+            this.stack = combined.toString();
+
+            // We update our message to a normalized one from the combined
+            // stacks, as this gives us a nice causal relationship.
+            this.message = combined.message;
+        }
```

Comment:
This is new art based on the old ways.
1. webapp would just swap the KAError stack trace for the source error stack trace
2. mobile would concat the stacks together, with the source error first and the KAError stack second
3. Now, in this version, we do some work to combine the stacks in a chronological manner that hopefully retains value but reduces repetition. See `ErrorInfo.combine` for details.